src/plugins/visualizations/public/vis.ts

```diff
@@ -147,22 +165,31 @@ export class Vis {
   }
 
   clone() {
-    return new Vis(this.type.name, this.serialize());
+    const { data, ...restOfSerialized } = this.serialize();
+    const vis = new Vis(this.type.name, restOfSerialized as any);
+    vis.setState(restOfSerialized as any);
+    const aggs = this.data.indexPattern
+      ? getAggs().createAggConfigs(this.data.indexPattern, data.aggs)
+      : undefined;
+    vis.data = {
+      ...this.data,
+      aggs,
+    };
+    return vis;
```

Comment:
I'm wondering if this could be simplified. Since you are relying on mutating `vis.data` after the fact anyway, why do you need to separate it from `restOfSerialized`?

Would this do the same thing?

```ts
    const serializedVis = this.serialize();
    const vis = new Vis(this.type.name, serializedVis);
    vis.setState(serializedVis);
    vis.data.aggs = this.data.indexPattern
      ? getAggs().createAggConfigs(this.data.indexPattern, serializedVis.data.aggs)
      : undefined;
    return vis;
```

Reply:
no, to keep clone method async we must not call setState with data property set.
we work around loading index pattern and linked search by keeping the same references.